[DOC] Add pipeline processor documentation (#5986)
opensearch-project/documentation-website: _ingest-pipelines/processors/pipeline.md

---
layout: default
title: Pipeline
parent: Ingest processors
nav_order: 200
---

# Pipeline processor

> The `pipeline` processor allows a pipeline to reference and include another predefined pipeline. This can be useful when you have a set of common processors that need to be shared across multiple pipelines. Instead of redefining those common processors in each pipeline, you can create a separate base pipeline containing the shared processors, and then reference that base pipeline from other pipelines using the pipeline processor.

Reviewer: ```suggestion
The `pipeline` processor allows a pipeline to reference and include another predefined pipeline. This can be useful when you have a set of common processors that need to be shared across multiple pipelines. Instead of redefining those common processors in each pipeline, you can create a separate base pipeline containing the shared processors and then reference that base pipeline from other pipelines using the pipeline processor.
```

The following is the syntax for the `pipeline` processor:

```json
{
  "pipeline": {
    "name": "general-pipeline"
  }
}
```
{% include copy-curl.html %}

## Configuration parameters

The following table lists the required and optional parameters for the `pipeline` processor.

Parameter | Required/Optional | Description |
|-----------|-----------|-----------|
`name` | Required	| The name of the pipeline to execute.
`description` | Optional | A description of the processor's purpose or configuration.
`if` | Optional | Specifies to conditionally execute the processor.

> `ignore_failure` | Optional | Specifies to ignore failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).

Reviewer: ```suggestion
`ignore_failure` | Optional | Specifies to ignore processor failures. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
```

> `on_failure` | Optional | Specifies to handle failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).

Reviewer: ```suggestion
`on_failure` | Optional | Specifies to handle processor failures. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
```

        "name": "general-pipeline"  
      }  
    }  
  ]  
}
```
{% include copy-curl.html %}

### Step 2 (Optional): Test the pipeline

It is recommended that you test your pipeline before you ingest documents.
{: .tip}

To test the pipeline, run the following query:

```json
POST _ingest/pipeline/outer-pipeline/_simulate
{  
  "docs": [  
    {  
      "_source": {  
        "protocol": "https",  
        "name":"test"  
      }  
    }  
  ]  
}  
```
{% include copy-curl.html %}

#### Response

The following example response confirms that the pipeline is working as expected:

```json
{  
  "docs": [  
    {  
      "doc": {  
        "_index": "_index",  
        "_id": "_id",  
        "_source": {  
          "protocol": "HTTPS"  
        },  
        "_ingest": {  
          "timestamp": "2024-05-24T02:43:43.700735801Z"  
        }  
      }  
    }  
  ]  
}
```
{% include copy-curl.html %}

### Step 3: Ingest a document 

The following query ingests a document into an index named `testindex1`:

```json
POST testindex1/_doc/1?pipeline=outer-pipeline  
{  
  "protocol": "https",  
  "name": "test"  
}  
```
{% include copy-curl.html %}

#### Response

> The request indexes the document with the `protocol` field converted to uppercase and the field name removed into the index `testindex1`, as shown in the following response:

Reviewer: "removed into" => either "moved to" or "removed from".

```json
{  
  "_index": "testindex1",  
  "_id": "1",  
  "_version": 2,  
  "result": "created",  
  "_shards": {  
    "total": 2,  
    "successful": 2,  
    "failed": 0  
  },  
  "_seq_no": 1,  
  "_primary_term": 1  
}  
```
{% include copy-curl.html %}

### Step 4 (Optional): Retrieve the document

To retrieve the document, run the following query:

```json
GET testindex1/_doc/1
```
{% include copy-curl.html %}

#### Response

> The response show the document with the `protocol` field converted to uppercase and the field name removed, as shown in the following response:

Reviewer: ```suggestion
The response shows the document with the `protocol` field converted to uppercase and the field name removed:
```

---
layout: default
title: Pipeline
parent: Ingest processors
nav_order: 200
---

# Pipeline processor

> The `pipeline` processor defines a series of processing steps to be executed on documents during the ingestion or update process. This is particularly useful when you need to perform complex data transformations or apply multiple processing steps to your data before it is indexed.

Reviewer 2: `pipeline` processor makes a pipeline can reference another pre-defined pipeline, it can be used when there's a general pipeline with some common processors which can be shared among other pipelines, users can reference the general pipeline directly without redefining the common processors.

The following is the syntax for the `pipeline` processor:

```json

> {

Reviewer 2: {
  "pipeline": {
    "name": "general-pipeline"
  }
}

  "pipeline": {
    "name": "<pipeline_name>",
    "description": "<pipeline_description>",
    "processors": [
      {
        "<processor_type>": {
          "<processor_config>": "<processor_value>"
        }
      },
      {
        "<processor_type>": {
          "<processor_config>": "<processor_value>"
        }
      }
    ]
  }
}
```
{% include copy-curl.html %}

## Configuration parameters

The following table lists the required and optional parameters for the `pipeline` processor.

Parameter | Required/Optional | Description |
|-----------|-----------|-----------|
`name` | Required	| The name of the pipeline to execute.

> `processors` | Required | An array of other processor configurations to execute in the specified order.

Reviewer 2: `processors` and `ignore_missing_pipeline ` should be removed, this processor doesn't have these parameters.

`ignore_missing_pipeline` | Optional | Determines whether to ignore a missing or undefined pipeline. Default is `false`.
`description` | Optional | Description of the processor's purpose or configuration.
`if` | Optional | Conditionally execute the processor.
`ignore_failure` | Optional | Ignore failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
`on_failure` | Optional | Handle failures for the processor. See [Handling pipeline failures]({{site.url}}{{site.baseurl}}/ingest-pipelines/pipeline-failures/).
`tag` | Optional | Identifier for the processor. Useful for debugging and metrics.

## Using the processor

Follow these steps to use the processor in a pipeline.

### Step 1: Create a pipeline

The following query creates a pipeline named `my_pipeline` that uses the `pipeline` processor to add a new field `my_field` with a static value and convert the `message` field to uppercase: 

```json

> PUT _ingest/pipeline/my_pipeline

Reviewer 2: The following query creates a general pipeline named `general-pipeline`, and then creates a new pipeline named `outer-pipeline` which references the `general-pipeline`.
Create a general pipeline:
```
PUT _ingest/pipeline/general_pipeline
{
  "description": "a general pipeline",
  "processors": [
    {
      "uppercase": {
        "field": "protocol"
      },
      "remove": {
        "field": "name"
      }
    }
  ]
}
```
Create an outer pipeline:
```
PUT _ingest/pipeline/outer-pipeline
{
  "description": "an outer pipeline referencing the general pipeline",
  "processors": [
    {
      "pipeline": {
        "name": "general-pipeline"
      }
    }
  ]
}
```

{
  "description": "My custom pipeline",
  "processors": [
    {
      "set": {
        "field": "my_field",
        "value": "static_value"
      }
    },
    {
      "uppercase": {
        "field": "message"
      }
    }
  ]
}

```
{% include copy-curl.html %}

### Step 2 (Optional): Test the pipeline

It is recommended that you test your pipeline before you ingest documents.
{: .tip}

To test the pipeline, run the following query:

```json

> POST _ingest/pipeline/my_pipeline/_simulate

Reviewer 2: ```
POST _ingest/pipeline/outer-pipeline/_simulate
{
  "docs": [
    {
      "_source": {
        "protocol": "https",
        "name":"test"
      }
    }
  ]
}
```

{
  "docs": [
    {
      "_source": {
        "message": "hello world"
      }
    }
  ]
}
```
{% include copy-curl.html %}

#### Response

The following example response confirms that the pipeline is working as expected:

> ```json

Reviewer 2: ```
{
  "docs": [
    {
      "doc": {
        "_index": "_index",
        "_id": "_id",
        "_source": {
          "protocol": "HTTPS"
        },
        "_ingest": {
          "timestamp": "2024-05-24T02:43:43.700735801Z"
        }
      }
    }
  ]
}
```

{
  "docs": [
    {
      "doc": {
        "_index": "_index",
        "_id": "_id",
        "_source": {
          "my_field": "static_value",
          "message": "HELLO WORLD"
        },
        "_ingest": {
          "timestamp": "2024-05-23T22:28:49.718036588Z"
        }
      }
    }
  ]
}
```
{% include copy-curl.html %}

### Step 3: Ingest a document 

> The following query ingests a document into an index named `my_index`:

Reviewer 2: The following query ingests a document into an index named `testindex1 `:

```json

> PUT my_index/_doc/1?pipeline=my_pipeline

Reviewer 2: ```
POST testindex1/_doc/1?pipeline=outer-pipeline
{
  "protocol": "https",
  "name": "test"
}
```

{
  "message": "hello world"
}
```
{% include copy-curl.html %}

#### Response

> The request indexes the document into the index `my_index` and then indexes all documents with the `message` field converted to uppercase and a new `my_field` field added with the value `static_value`.

Reviewer 2: The request indexes the document with the `protocol` field converted to uppercase and the field `name` removed into the index `testindex1 `.

```json

> {

Reviewer 2: ```
{
  "_index": "testindex1",
  "_id": "1",
  "_version": 2,
  "result": "created",
  "_shards": {
    "total": 2,
    "successful": 2,
    "failed": 0
  },
  "_seq_no": 1,
  "_primary_term": 1
}
```

  "_index": "my_index",
  "_id": "1",
  "_version": 1,
  "result": "created",
  "_shards": {
    "total": 2,
    "successful": 1,
    "failed": 0
  },
  "_seq_no": 0,
  "_primary_term": 1
}
```
{% include copy-curl.html %}

### Step 4 (Optional): Retrieve the document

To retrieve the document, run the following query:

```json

> GET my_index/_doc/1

Reviewer 2: GET testindex1/_doc/1

```
{% include copy-curl.html %}

#### Response

> The response shows the document with the `message` field converted to uppercase and the `my_field` field added with the value `static_value`.

Reviewer 2: The response show the document with the protocol field converted to uppercase and the field name removed.

> ```json

Reviewer 2: ```
{
  "_index": "testindex1",
  "_id": "1",
  "_version": 2,
  "_seq_no": 1,
  "_primary_term": 1,
  "found": true,
  "_source": {
    "protocol": "HTTPS"
  }
}
```
